packages/opentelemetry-tracer-web/src/StackScopeManager.ts

```diff
@@ -0,0 +1,161 @@
+/*!
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ScopeManager } from '@opentelemetry/scope-base';
+
+export class StackScopeManager implements ScopeManager {
```

Comment:
Could you add some documentation for this class about how it works? In particular, to highlight its limitations in terms of not keeping context across async calls?

Related to that, what's the value of the StackScopeManager here over a simpler model where it just stores a single scope at a time? I think storing a single scope will work for basic web cases such as the initial load or even user interactions where we assume only one thing is happening at a time. 

For concurrent user interactions, I think we really need something that is async-aware like a Zone.js based scope manager - but that's a more advanced use case.

Reply:
With regards to value of StackScopeManager over simpler model. Can you please provide what you mean by simpler model ?. The BasicTracer expects to define a scopeManager with certain functionality:
https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-basic-tracer/src/types.ts#L63.
So I would say that StackScopeManager is exactly the simpler model , unless I'm missing something. The more advanced would be the one that will cover the async functionality as you said.

Reply:
What if you had a ScopeManager implementation that looked like this:

```
export interface ScopeManager {
   private _scope: unknown = {};

  active(): unknown {
    return this._scope;
  }

  with<T extends (...args: unknown[]) => ReturnType<T>>(
    scope: unknown,
    fn: T
  ): ReturnType<T> {
     this._scope = scope;
     return fn.apply(args);
  }

  bind<T>(target: T, scope?: unknown): T {
     if (scope !== undefined) {
       this._scope = scope;
     }
     return target;
  }
}
```

That to me feels like the most minimal scope manager that actually does something: it stores the scope that you give it in a single variable and makes no attempt to propagate it or make it hierarchical or async in any way. (Of course it's missing some things like documentation, error handling, etc. but gives the idea)

That would be sufficient for some browser use cases (e.g. the initial document load) where we really just have one global scope.

Sorry for the slow response on this! I think I missed your reply an earlier time I looked at this.

Reply:
Alternatively, here is a slightly more complex version of `with` that captures the scope and restores it without needing an explicit data structure (could do the same for `bind`):

```
 with<T extends (...args: unknown[]) => ReturnType<T>>(
    scope: unknown,
    fn: T
  ): ReturnType<T> {
     return () => {
        const oldScope = this._scope;
        this._scope = scope;
        return fn.apply(args);
        this._scope = oldScope;
     };
  }
```
This still feels a little lighter weight, but I think I would prefer the most simple version where we just have one scope at a time explicitly since we don't do any async scope propagation, and then have the Zone.js based one do the fancy propagation and hierarchical wrapping.

Reply:
@draffensperger ok I have simplified the scope manager. Previously I had identified the scopes by adding some of ids to them. It was also to be able to identify them later in async but that is not the case anymore as we will rather use the new scope manager for that something like `ZoneScopeManager`. I have also realised that there might be multiple tracer so adding id to scope is not the best idea. Keeping it shortly then I would leave the StackScopeManager and treat it as really the basic one as in your example there would be a simple problem with things like calling the callback with scope inside another callback - it would not be restored after all.

Reply:
Yeah, that makes sense to want to make sure we can at least restore scope for simple callbacks.

Is there a way you can use closures (function wrapping via arrow functions) to restore the scope rather than using UIDs and an object? I think that might be a little bit cleaner and use less code, but maybe there's a use case I'm not understanding here

Reply:
@draffensperger extra closure is not needed in this case , the `fn` is a callback function that is called immediately with the current scope, there is no waiting for that like with `bind` function. Unless I'm not aware of something.

Reply:
Ah, I see. I guess then `bind` would need storage via the closure, and `with` could just store it in a local variable?

Between those techniques is it possible to change `_scopesStack` to just an `_scope` variable and simplify the logic?

Reply:
@draffensperger yes I was just referring to the extra closure as I wasn't sure if I'm not missing anything, the rest is fine, just pushed the changes you can have a look, thx

Reply:
Thanks for the refactoring! I will make a few more comments, but they are more stylistic. I think this is really coming together well!